File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/java/com/divideconquer/algorithms/ExponentiationBySquaring.java
package com.divideconquer.algorithms;

public final class ExponentiationBySquaring {
    private ExponentiationBySquaring() {
    }

    public static long pow(long base, int exponent) {
        if (exponent < 0) {
            throw new IllegalArgumentException("exponent must be non-negative");
        }
        if (exponent == 0) {
            return 1;
        }
        long half = pow(base, exponent / 2);
        long squared = half * half;
        return exponent % 2 == 0 ? squared : squared * base;
    }
}

File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/python/divide_conquer/algorithms/exponentiation_by_squaring.py
"""Exponentiation by squaring."""


def pow_int(base: int, exponent: int) -> int:
    """Return base raised to a non-negative integer exponent."""
    if exponent < 0:
        raise ValueError("exponent must be non-negative")
    if exponent == 0:
        return 1

    half = pow_int(base, exponent // 2)
    squared = half * half
    return squared if exponent % 2 == 0 else squared * base

